### PR TITLE
Refactor Mail Sender to allow multiple recipients

### PR DESCRIPTION
**Pull request by MiguelAHM**

### CODE REVIEW

File: whois-api/src/test/resources/testMail/permanentFailureMessageRfc822MultipleRecipients.mail [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+MIME-Version: 1.0
+Content-Type: multipart/report; report-type=delivery-status; boundary="6KTYe.5maleLdVv.5/WobWlHy4k.7vnm51B"
+Delivered-To: test-dbm@ripe.net
+
+--6KTYe.5maleLdVv.5/WobWlHy4k.7vnm51B
+content-type: text/plain;
+    charset="utf-8"
+Content-Transfer-Encoding: quoted-printable
+
+The following message to <nonexistant@host.org> was undeliverable.
```
Comment (eshryane) [inline whois-api/src/test/resources/testMail/permanentFailureMessageRfc822MultipleRecipients.mail:30]: Does this mean that *only* `nonexistant@ripe.net` was undeliverable? It's also possible that multiple recipients to the same mail provider are all undeliverable

File: whois-api/src/test/java/net/ripe/db/whois/api/mail/dequeue/BouncedMessageParserTest.java [hunk 2]
```diff
@@ -27,23 +28,32 @@ public void parse_permanent_delivery_failure_message_rfc822() throws Exception {
...
-        assertThat(bouncedMessage.emailAddress(), is("nonexistant@host.org"));
+        assertThat(bouncedMessage.emailAddresses().get(0), is("nonexistant@host.org"));
+    }
+
+    @Test
+    public void parse_permanent_delivery_failure_multiple_recipients_rfc822() throws Exception {
+        final MessageInfo bouncedMessage = subject.parse(MimeMessageProvider.getUpdateMessage("permanentFailureMessageRfc822MultipleRecipients.mail"));
+
+        assertThat(bouncedMessage.messageId(), is("XXXXXXXX-5AE3-4C58-8E3F-860327BA955D@ripe.net"));
+        assertThat(bouncedMessage.emailAddresses(), containsInAnyOrder("nonexistant@host.org", "nonexistant1@host.org"));
```
Comment (eshryane) [inline whois-api/src/test/java/net/ripe/db/whois/api/mail/dequeue/BouncedMessageParserTest.java:39]: How are *both* addresses bounced? I only see in `permanentFailureMessageRfc822MultipleRecipients.mail` that `nonexistant@host.org` bounced, not `nonexistant1@host.org`

Comment (MiguelAHM) [reply whois-api/src/test/java/net/ripe/db/whois/api/mail/dequeue/BouncedMessageParserTest.java:39]: both appear in the rfc822
To: nonexistant@host.org, nonexistant1@host.org

probably this use case is incomplete, we need a failure per each

Comment (MiguelAHM) [reply whois-api/src/test/java/net/ripe/db/whois/api/mail/dequeue/BouncedMessageParserTest.java:48]: ok, added

File: whois-api/src/test/resources/testMail/permanentFailureMessageRfc822MultipleRecipients.mail (hunk 1 above)
Comment (MiguelAHM) [reply whois-api/src/test/resources/testMail/permanentFailureMessageRfc822MultipleRecipients.mail]: If we have 10 recipients in one email and 5 of them are bounced
We will receive a report for those 5 emails, not for all of 10

In this test both are undeliverable but we just received one just report. This should happen, we should receive a report per recipient...but it is good to have as we discussed

File: whois-api/src/test/java/net/ripe/db/whois/api/mail/dequeue/MessageServiceTestIntegration.java [hunk 3]
```diff
@@ -90,6 +90,22 @@ public void testBouncedEmailFromCorrectEmail() throws MessagingException, IOExce
         assertThat(isUndeliverableAddress(BOUNCED_MAIL_RECIPIENT), is(true));
     }
 
+    @Test
+    public void testBouncedMultipleEmailFromCorrectEmail() throws MessagingException, IOException {
+        insertOutgoingMessageId("XXXXXXXX-5AE3-4C58-8E3F-860327BA955D@ripe.net", BOUNCED_MAIL_RECIPIENT);
+        insertOutgoingMessageId("XXXXXXXX-5AE3-4C58-8E3F-860327BA955D@ripe.net", ANOTHER_BOUNCED_MAIL_RECIPIENT);
```
[the remark this answers is not in the document]
Comment (MiguelAHM) [reply whois-api/src/test/java/net/ripe/db/whois/api/mail/dequeue/MessageServiceTestIntegration.java:96]: yes, we check that the recipients in the email that we receive exists in the database for that specific message_id.
If not anyone can send us any message with the recipients that they want and we can be setting those recipients as unsubscribed by mistake (because that "anyone" sent to us a fake mail)
So we are checking that the message_id exist and that the recipients of that message exist in our database for that specific message Id

[the remark this answers is not in the document]
Comment (MiguelAHM) [reply whois-api/src/test/java/net/ripe/db/whois/api/mail/dequeue/MessageServiceTestIntegration.java:96]: so I need to add this information in the outgoing table for bypass this check

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/BouncedMessageParser.java [hunk 4]
```diff
@@ -33,29 +35,24 @@ public BouncedMessageParser(@Value("${mail.smtp.from:}") final String smtpFrom)
...
+                        final MimeMessage returnedMessage = multipartReport.getReturnedMessage();
+                        final String messageId = getMessageId(returnedMessage.getMessageID());
+                        final List<String> recipient = extractRecipients(deliveryStatus);
+                        return new MessageInfo(recipient, messageId);
+                    }
                 }
             }
+            return null;
+        } catch (MessagingException | IOException ex){
+            throw new ParseException();
```
Comment (MiguelAHM) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/BouncedMessageParser.java:54]: pass the ex into the ParseException

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/BouncedMessageParser.java [hunk 5]
```diff
@@ -84,6 +81,19 @@ private DeliveryStatus deliveryStatus(final Part part) throws MessagingException
...
     }
 
+    private List<String> extractRecipients(final DeliveryStatus deliveryStatus) {
+        final List<String> recipients = Lists.newArrayList();
+        for (int dsn = 0; dsn < deliveryStatus.getRecipientDSNCount(); dsn++) {
+            final String recipient = getHeaderValue(deliveryStatus.getRecipientDSN(dsn), "Final-Recipient");
+            if (recipient == null){
+                continue;
+            }
+            final String extractedAddress = recipient.substring("rfc822".length() + 1).trim();
```
Comment (MiguelAHM) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/BouncedMessageParser.java:91]: use a regex instead
log an error if no match

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageDequeue.java [hunk 6]
```diff
@@ -213,6 +214,8 @@ private void handleMessage(final String messageId) {
                 loggerContext.remove();
             }
 
+        } catch (ParseException e){
```
Comment (MiguelAHM) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageDequeue.java:217]: remove it now

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java [hunk 7]
```diff
@@ -37,43 +36,48 @@ public MessageService(
...
+        final List<String> emails = outgoingMessageDao.getEmails(message.messageId());
 
-        if (isIncorrectMessage(message, email)){
+        if (isIncorrectMessage(message, emails)){
             return;
         }
 
-        LOGGER.debug("Undeliverable message-id {} email {}", message.messageId(), message.emailAddress());
-        emailStatusDao.createEmailStatus(email, EmailStatus.UNDELIVERABLE);
+        LOGGER.debug("Undeliverable message-id {} email {}", message.messageId(), message.emailAddresses());
```
Comment (MiguelAHM) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:53]: does log list properly? (use a joiner if not)

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java [hunk 8]
```diff
@@ -37,43 +36,48 @@ public MessageService(
...
+        }
+
+        if (emails == null || emails.isEmpty()) {
             LOGGER.warn("Couldn't find outgoing message matching {}", message.messageId());
             return true;
         }
 
-        if (!email.equalsIgnoreCase(message.emailAddress())) {
-            LOGGER.warn("Email {} in outgoing message doesn't match '{}' in failure response", email, message.emailAddress());
+        if (!emails.containsAll(message.emailAddresses())) {
```
Comment (MiguelAHM) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:79]: this is not case sensitive 
anyMatch of stream

File: whois-api/src/test/java/net/ripe/db/whois/api/mail/dequeue/BouncedMessageParserTest.java [hunk 9]
```diff
@@ -27,23 +28,41 @@ public void parse_permanent_delivery_failure_message_rfc822() throws Exception {
...
     }
 
     @Test
     public void parse_permanent_delivery_failure_rfc822_headers() throws Exception {
         final MessageInfo bouncedMessage = subject.parse(MimeMessageProvider.getUpdateMessage("permanentFailureRfc822Headers.mail"));
 
         assertThat(bouncedMessage.messageId(), is("XXXXXXXX-2BCC-4B29-9D86-3B8C68DD835D@ripe.net"));
-        assertThat(bouncedMessage.emailAddress(), is("nonexistant@ripe.net"));
+        assertThat(bouncedMessage.emailAddresses().get(0), is("nonexistant@ripe.net"));
```
Comment (MiguelAHM) [inline whois-api/src/test/java/net/ripe/db/whois/api/mail/dequeue/BouncedMessageParserTest.java:56]: better use contains in any order for both (before)

File: whois-api/src/test/java/net/ripe/db/whois/api/mail/dequeue/UnsubscribeMessageParserTest.java [hunk 10]
```diff
@@ -23,15 +23,15 @@ public void unsubscribe_from_apple_mail() throws Exception {
         final MessageInfo unsubscribedMessage = subject.parse(MimeMessageProvider.getUpdateMessage("unsubscribeAppleMail.mail"));
 
         assertThat(unsubscribedMessage.messageId(), is("8b8ed6c0-f9cc-4a5f-afbb-fde079b94f44@ripe.net"));
-        assertThat(unsubscribedMessage.emailAddress(), is("enduser@ripe.net"));
+        assertThat(unsubscribedMessage.emailAddresses().get(0), is("enduser@ripe.net"));
```
Comment (MiguelAHM) [inline whois-api/src/test/java/net/ripe/db/whois/api/mail/dequeue/UnsubscribeMessageParserTest.java:26]: contains in any order

File: whois-commons/src/main/java/net/ripe/db/whois/common/dao/OutgoingMessageDao.java [hunk 11]
```diff
@@ -26,15 +29,15 @@ public void saveOutGoingMessageId(final String uuid, final String email){
     }
 
     @Nullable
```
Comment (MiguelAHM) [inline whois-commons/src/main/java/net/ripe/db/whois/common/dao/OutgoingMessageDao.java:31]: Check if we can remove nullable is not results

File: whois-commons/src/main/resources/internals_schema.sql [hunk 12]
```diff
@@ -174,7 +174,7 @@ CREATE TABLE `outgoing_message` (
    `message_id` varchar(80) NOT NULL,
    `email` varchar(80) NOT NULL,
    `last_update` datetime DEFAULT now(),
-   PRIMARY KEY (`message_id`)
+   PRIMARY KEY (`message_id`, `email`)
```
Comment (MiguelAHM) [inline whois-commons/src/main/resources/internals_schema.sql:177]: Remember to deploy this change

File: whois-update/src/main/java/net/ripe/db/whois/update/mail/WhoisMailGatewaySmtp.java [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+import jakarta.mail.MessagingException;
+import net.ripe.db.whois.common.Message;
+import net.ripe.db.whois.common.Messages;
+import net.ripe.db.whois.common.aspects.RetryFor;
+import net.ripe.db.whois.common.dao.EmailStatusDao;
+import net.ripe.db.whois.common.dao.OutgoingMessageDao;
+import net.ripe.db.whois.common.mail.EmailStatus;
+import net.ripe.db.whois.update.domain.ResponseMessage;
+import net.ripe.db.whois.update.log.LoggerContext;
+import org.apache.commons.lang.NotImplementedException;
```
Comment (MiguelAHM) [inline whois-update/src/main/java/net/ripe/db/whois/update/mail/WhoisMailGatewaySmtp.java:12]: USE
UnsupportedOperationException

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 7 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:54]: I need to check all the lists and add a joiner (already in my agenda)

File: whois-update/src/main/java/net/ripe/db/whois/update/mail/MailGateway.java [hunk 14]
```diff
@@ -2,8 +2,13 @@
 
 import net.ripe.db.whois.update.domain.ResponseMessage;
 
+import java.util.Set;
+
 
 public interface MailGateway {
```
Comment (maggarwal13) [inline whois-update/src/main/java/net/ripe/db/whois/update/mail/MailGateway.java:8]: There is no need for interface if there is an abstract class?

File: whois-commons/src/main/resources/patch/internals-1.111.sql [hunk 15]
```diff
@@ -5,7 +5,7 @@ CREATE TABLE `outgoing_message` (
      `message_id` varchar(80) NOT NULL,
      `email` varchar(80) NOT NULL,
      `last_update` datetime DEFAULT now(),
-     PRIMARY KEY (`message_id`)
+     PRIMARY KEY (`message_id`, `email`)
```
Comment (maggarwal13) [inline whois-commons/src/main/resources/patch/internals-1.111.sql:8]: Why this change requires email column?

Comment (MiguelAHM) [reply whois-commons/src/main/resources/patch/internals-1.111.sql:8]: 1 mail can have multiple recipients. So we can have two records with same message_id

Comment (maggarwal13) [reply whois-commons/src/main/resources/patch/internals-1.111.sql:8]: Sorry I was not expecting so many changes for this? Seems like a seperate PR which says support multiple recipients as it needs changes in Dao layer and then a seperate PR for this?

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/MessageInfo.java [hunk 16]
```diff
@@ -1,8 +1,9 @@
 package net.ripe.db.whois.api.mail;
 
 import javax.annotation.Nullable;
+import java.util.List;
 
-public record MessageInfo(String emailAddress, String messageId) {
+public record MessageInfo(List<String> emailAddresses, String messageId) {
```
Comment (MiguelAHM) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/MessageInfo.java:6]: EmailMessageInfo

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java [hunk 17]
```diff
@@ -40,43 +41,60 @@ public MessageService(
...
 
-        LOGGER.debug("Unsubscribe message-id {} email {}", message.messageId(), message.emailAddress());
-        emailStatusDao.createEmailStatus(email, EmailStatus.UNSUBSCRIBE);
+        LOGGER.debug("Unsubscribe message-id {} email {}", message.messageId(), StringUtils.join(message.emailAddresses(), ", "));
+        emails.forEach(email -> emailStatusDao.createEmailStatus(email, EmailStatus.UNSUBSCRIBE));
     }
 
-    private boolean isIncorrectMessage(final MessageInfo message, final String email){
-        if (Strings.isNullOrEmpty(email)) {
+    private boolean isIncorrectMessage(final MessageInfo message, final List<String> emails){
```
Comment (maggarwal13) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:70]: Can you rename to isValidEmailMessage

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java [hunk 18]
```diff
@@ -40,43 +41,60 @@ public MessageService(
...
         }
 
-        LOGGER.debug("Undeliverable message-id {} email {}", message.messageId(), message.emailAddress());
-        emailStatusDao.createEmailStatus(email, EmailStatus.UNDELIVERABLE);
+        LOGGER.debug("Undeliverable message-id {} email {}", message.messageId(), StringUtils.join(message.emailAddresses(), ", "));
+        message.emailAddresses().forEach(email -> emailStatusDao.createEmailStatus(email, EmailStatus.UNDELIVERABLE));
     }
 
     public void verifyAndSetAsUnsubscribed(final MessageInfo message){
-        final String email = outgoingMessageDao.getEmail(message.messageId());
```
Comment (maggarwal13) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:59]: when you unsubscribe you cannot get multiple emails for a single unsubscribe request?

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java [hunk 19]
```diff
@@ -40,43 +41,60 @@ public MessageService(
...
             return;
         }
 
-        LOGGER.debug("Undeliverable message-id {} email {}", message.messageId(), message.emailAddress());
-        emailStatusDao.createEmailStatus(email, EmailStatus.UNDELIVERABLE);
+        LOGGER.debug("Undeliverable message-id {} email {}", message.messageId(), StringUtils.join(message.emailAddresses(), ", "));
+        message.emailAddresses().forEach(email -> emailStatusDao.createEmailStatus(email, EmailStatus.UNDELIVERABLE));
     }
 
     public void verifyAndSetAsUnsubscribed(final MessageInfo message){
```
Comment (maggarwal13) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:59]: (messageId, email )?

File: whois-update/src/main/java/net/ripe/db/whois/update/mail/WhoisMailGatewaySmtp.java [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    @Override
+    public void sendEmail(final String to, final ResponseMessage responseMessage) {
+        sendEmail(to, responseMessage.getSubject(), responseMessage.getMessage(), responseMessage.getReplyTo());
+    }
+
+    @Override
+    public void sendEmail(final String to, final String subject, final String text, @Nullable final String replyTo) {
+        if (!mailConfiguration.isEnabled()) {
```
Comment (maggarwal13) [inline whois-update/src/main/java/net/ripe/db/whois/update/mail/WhoisMailGatewaySmtp.java:54]: Is this should be in abstract class?

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java [hunk 21]
```diff
@@ -40,43 +41,60 @@ public MessageService(
     public MessageInfo getBouncedMessageInfo(final MimeMessage message) throws MessagingException, IOException {
         return bouncedMessageParser.parse(message);
     }
-    public MessageInfo getUnsubscribedMessageInfo(final MimeMessage message) throws MessagingException, IOException {
+    public MessageInfo getUnsubscribedMessageInfo(final MimeMessage message) throws MessagingException {
         return unsubscribeMessageParser.parse(message);
     }
 
     public void verifyAndSetAsUndeliverable(final MessageInfo message){
-        final String email = outgoingMessageDao.getEmail(message.messageId());
```
[the remark this answers is not in the document]
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:48]: I put outgoingEmails

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 18 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:59]: yes, I still want to keep the method signature as it is. 
The difference is just that the message.emailAddresses() is just one. So I prefer to check that this is true and log an error

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/EmailMessageInfo.java [hunk 22]
```diff
@@ -12,7 +13,7 @@ public String messageId() {
 
     @Override
     @Nullable
-    public String emailAddress() {
-        return emailAddress;
+    public List<String> emailAddresses() {
```
Comment (maggarwal13) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/EmailMessageInfo.java:16]: You do not need a getter method for a record?

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java [hunk 23]
```diff
@@ -37,46 +38,68 @@ public MessageService(
...
-            LOGGER.warn("Email {} in outgoing message doesn't match '{}' in failure response", email, message.emailAddress());
+        if (!containsAllCaseInsensitive(message.emailAddresses(), outgoingEmail)) {
+            LOGGER.warn("Email {} in outgoing message doesn't match '{}' in failure response", outgoingEmail, StringUtils.join(message.emailAddresses(), ", "));
             return true;
         }
         return false;
     }
 
+    private boolean containsAllCaseInsensitive(final List<String> messageRecipients, final List<String> storedEmails){
+        final List<String> emailsInLowerCase = storedEmails
```
Comment (maggarwal13) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:94]: It will always be false specially incase of unsubscribe? It should be contains rather then matchAll?

File: whois-commons/src/main/java/net/ripe/db/whois/common/dao/OutgoingMessageDao.java [hunk 24]
```diff
@@ -25,16 +27,15 @@ public void saveOutGoingMessageId(final String uuid, final String email){
...
-            "SELECT email from outgoing_message where message_id = ?",
-            new Object[] { messageId },
-            (ResultSet resultSet) -> {
-                if (resultSet.next()) {
-                    return resultSet.getString(1);
-                }
-                return null;
-            });
+                "SELECT email from outgoing_message where message_id = ?",
+                new RowMapper<String>() {
```
Comment (maggarwal13) [inline whois-commons/src/main/java/net/ripe/db/whois/common/dao/OutgoingMessageDao.java:34]: use queryFoList instead of query and a rowmapper

File: whois-update/src/main/java/net/ripe/db/whois/update/mail/MailGatewaySmtp.java [hunk 25]
```diff
@@ -5,84 +5,90 @@
...
+    @Nullable
+    protected String extractEmailBetweenAngleBrackets(final String email) {
+        if(email == null) {
+            return null;
+        }
+
+        try {
+            return new InternetAddress(email).getAddress();
+        } catch (AddressException e) {
+            return email;
```
Comment (maggarwal13) [inline whois-update/src/main/java/net/ripe/db/whois/update/mail/MailGatewaySmtp.java:60]: In what cases it can throw AddressException ? I think if should cehck for angle brackets only thn try to parse???

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 23 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:94]: messageRecipients -> test@ripe.net, test1@ripe.net
emailsInLowerCase (email in outgoing table) -> test@ripe.net, test1@ripe.net, test2@ripe.net

test2@ripe.net -> succeeded

loop:
(test@ripe.net, test1@ripe.net).allMatch((test@ripe.net, test1@ripe.net, test2@ripe.net).contains(test@ripe.net))
(test@ripe.net, test1@ripe.net).allMatch((test@ripe.net, test1@ripe.net, test2@ripe.net).contains(test1@ripe.net))

It is fine, right? In case if unsubscribe just modify messageRecipients

File: whois-update/src/main/java/net/ripe/db/whois/update/mail/MailGatewaySmtp.java (hunk 25 above)
Comment (MiguelAHM) [reply whois-update/src/main/java/net/ripe/db/whois/update/mail/MailGatewaySmtp.java:60]: When is not able to parse the address seems.

    public InternetAddress(String address) throws AddressException {
        InternetAddress[] a = parse(address, true);
        if (a.length != 1) {
            throw new AddressException("Illegal address", address);
        } else {
            this.address = a[0].address;
            this.personal = a[0].personal;
            this.encodedPersonal = a[0].encodedPersonal;
        }
    }

This is already in master. This change was suggested by Ed. Using that library instead of just removing the angle brackets

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 23 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:94]: and the tests are green with this

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 19 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java]: (I will close it because already discussed in another comment)

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/EmailMessageInfo.java (hunk 22 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/EmailMessageInfo.java:16]: you, with a record you can get to that value without a get!

recordClass.propertyName

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java [hunk 26]
```diff
@@ -37,46 +38,68 @@ public MessageService(
...
-        emailStatusDao.createEmailStatus(email, EmailStatus.UNDELIVERABLE);
+        LOGGER.debug("Undeliverable message-id {} email {}", message.messageId(), StringUtils.join(message.emailAddresses(), ", "));
+        message.emailAddresses().forEach(email -> emailStatusDao.createEmailStatus(email, EmailStatus.UNDELIVERABLE));
     }
 
-    public void verifyAndSetAsUnsubscribed(final MessageInfo message){
-        final String email = outgoingMessageDao.getEmail(message.messageId());
+    public void verifyAndSetAsUnsubscribed(final EmailMessageInfo message){
+        if (message.emailAddresses() != null && message.emailAddresses().size() > 1){
+            LOGGER.error("This can not happen, unsubscribe with multiple recipients");
```
Comment (maggarwal13) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:61]: Warning would be better to avoid so many sentry alerts

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java [hunk 27]
```diff
@@ -37,46 +38,68 @@ public MessageService(
...
-        LOGGER.debug("Undeliverable message-id {} email {}", message.messageId(), message.emailAddress());
-        emailStatusDao.createEmailStatus(email, EmailStatus.UNDELIVERABLE);
+        LOGGER.debug("Undeliverable message-id {} email {}", message.messageId(), StringUtils.join(message.emailAddresses(), ", "));
+        message.emailAddresses().forEach(email -> emailStatusDao.createEmailStatus(email, EmailStatus.UNDELIVERABLE));
     }
 
-    public void verifyAndSetAsUnsubscribed(final MessageInfo message){
-        final String email = outgoingMessageDao.getEmail(message.messageId());
+    public void verifyAndSetAsUnsubscribed(final EmailMessageInfo message){
+        if (message.emailAddresses() != null && message.emailAddresses().size() > 1){
```
[the remark this answers is not in the document]
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:60]: yes, I was thinking the same. Perfect

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java [hunk 28]
```diff
@@ -37,46 +38,68 @@ public MessageService(
...
+            LOGGER.error("This can not happen, unsubscribe with multiple recipients");
+            return;
+        }
+
+        final List<String> outgoingEmail = outgoingMessageDao.getEmails(message.messageId());
```
Comment (maggarwal13) [inline whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:66]: There could be thousands of emails better to create a method isEmailExists(message_id , email ) in dao layer. But I do not see lots of unsubscribe requests so may be feel free to ignore

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/EmailMessageInfo.java (hunk 22 above)
Comment (maggarwal13) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/EmailMessageInfo.java:16]: yes remove the getters then ?

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 28 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:66]: In that case we need to call isEmailExists a lot of times, one per email
I prefer to keep that logic here and not in the ado

Comment (maggarwal13) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:66]: unsubscribe can opnly have one email

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/EmailMessageInfo.java (hunk 22 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/EmailMessageInfo.java:16]: ....true

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 28 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:66]: yes, sorry I was thinking for Undeliverable as well.
I can change that for unsubscribe yes

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 26 above)
Comment (maggarwal13) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:61]: Warnings are also reported in sentry. Error has higher priority so better keep at warning.

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 28 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java]: Done

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/EmailMessageInfo.java (hunk 22 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/EmailMessageInfo.java]: Done

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 17 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java]: Done

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java:48]: are you ok with this name?

File: whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java (hunk 26 above)
Comment (MiguelAHM) [reply whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/MessageService.java]: Done

**Review by maggarwal13 — APPROVED** [review]
